text/0000-interrupt-cc.md

```diff
@@ -0,0 +1,147 @@
+(ISRs) to implement certain functionality. In some cases this is a way to
+improve the system's efficiency, while in others it is required for correct
+operation.
+
+Most bare-metal software (such as a program designed to run on an ARM
+microcontroller) contains at least one custom-written ISR, supporting important
+system functionality. Oftentimes this is done in the interest of performance, so
+providing fast interrupt entry and exit in software is an important goal.
+
+In Rust today, the only option for implementation of such an ISR is to build the
+ISR entry point as assembly which may call into Rust code. For example,
+[RustOS][rustos], a simple x86 OS kernel implemented in Rust, currently
+implements ISRs as callbacks to a single "master" interrupt handler:
+
+[rustos]: https://github.com/ryanra/RustOS
+
+```
+.macro make_callback num
+  callback_\num\():
+.endm
+
+.macro make_all_callbacks, num=50
+.if \num+1
+   make_callback %num 
+      pushal
+      pushl $\num
+      call unified_handler
+      
+      addl $4, %esp
+      popal
+      iret
+  make_all_callbacks \num-1
+.endif
+.endm
+make_all_callbacks
+```
+
+This pair of assembler macros generate 50 different ISRs each of which simply
+call the master handler with the interrupt vector number as a parameter, which
+in turn dispatches to code to actually handle the interrupt. This kind of double
+dispatch, while functional, is extremely inefficient in all of code size (even
+unused ISRs must be generated), runtime memory usage (the double dispatch for
+every interrupt requires a minimum of 12 additional bytes on the stack), and
+speed (many ISRs are relatively simple, and the overhead of double dispatch may
+have a significant effect on performance).
+
+This kind of pattern is common in current bare-metal Rust programs: because the
+language provides no satisfactory way to approach the need for ISRs, programmers
+resort to inefficient ad-hoc solutions, which are usable but unlikely to be
+competitive with other systems programming languages. By comparison, exposing
+this information to the compiler both eliminates these inefficiencies and
+exposes more information to the compiler's optimizer for even greater potential
+gains.
+
+Assembly is required for this task because typically hardware interrupts require
+that the ISR preserve the values of all registers. Failure to do so will lead to
+unpredictable (inconsistent, even) behavior of the interrupted code. On some
+architectures, interrupt exit also requires a special instruction, such as
+`iret` on x86.
+
+# Detailed design
+
+Add a family of `interrupt` calling conventions to Rust, one for each
+architecture supported by the compiler. Given the current set of machines
+supported officially (the `arch` field in target specifications), they are as
+follow:
+
+ * `aarch64`
+ * `arm`
+ * `mips`
+ * `powerpc`
+ * `x86`
+ * `x86_64`
+
+Each of these architectures combines with a `_interrupt` suffix to form the full
```

Comment:
This came up in discussion too; I like the idea of having a single default `interrupt` CC available, but there needs to be at least a little more flexibility. Some systems have multiple interrupt CCs, such as some (older) ARMs where there are 'fast' and 'slow' interrupts which differ in what resources the hardware will automatically preserve.

We could certainly add a `"interrupt"` CC with defined defaults for each architecture, but that alone is not always sufficiently expressive. Some of the architecture prefixes could be done away with, but at that point I think it's appropriate to have the architecture-prefixed versions available for all systems regardless of which ones have multiple variants.

---

I suppose some discussion of that (and possible future expansion within the interrupt families) is in order here.